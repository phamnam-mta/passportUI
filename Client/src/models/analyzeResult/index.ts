import { V3_0_3_AnalyzeResult } from "./v3_0_3";

export type AnalyzeResponse = {
    createdDateTime: string;
    lastUpdatedDateTime: string;
    status: string;
    analyzeResult: V3_0_3_AnalyzeResult;
};

export type StudioAnalyzeResult = {
    apiVersion: string;
    modelId: string;
    stringIndexType: StringIndexType;
    content: string;
    pages: StudioDocumentPage[];
    blocks?: StudioBlock[];
    paragraphs?: StudioParagraph[];
    tables?: StudioDocumentTable[];
    keyValuePairs?: StudioDocumentKeyValuePair[];
    entities?: StudioDocumentEntity[];
    styles?: StudioDocumentStyle[];
    documents?: StudioDocumentResult[];
};

type StringIndexType = "textElements" | "unicodeCodePoint" | "utf16CodeUnit";

export type StudioDocumentPage = {
    pageNumber: number;
    angle: number;
    width: number;
    height: number;
    unit: "pixel" | "inch";
    words: StudioWord[];
    images?: StudioImage[];
    selectionMarks?: StudioSelectionMark[];
    lines?: StudioLine[];
    spans?: StudioSpan[];
    kind?: string;
};

export type StudioDocumentTable = {
    rowCount: number;
    columnCount: number;
    cells: StudioDocumentTableCell[];
    boundingRegions: BoundingRegion[];
    spans: StudioSpan[];
};

export type StudioDocumentKeyValuePair = {
    key: StudioKeyValuePairElement;
    value?: StudioKeyValuePairElement;
    confidence: Confidence;
};

export type StudioDocumentEntity = {
    category: string;
    subCategory?: string;
    content: string;
    boundingRegions: BoundingRegion[];
    confidence: Confidence;
    spans: StudioSpan[];
};

export type StudioDocumentStyle = {
    confidence?: Confidence;
    spans: StudioSpan[];
    isHandwritten?: boolean;
};

export type StudioDocumentResult = {
    docType: string;
    boundingRegions: BoundingRegion[];
    fields: { [fieldName: string]: StudioFieldValue };
    confidence: Confidence;
    spans: StudioSpan[];
};

export type BoundingRegion = {
    pageNumber: number;
    polygon: Polygon;
};

export type StudioWord = {
    content: string;
    polygon: Polygon;
    confidence?: Confidence;
    span?: StudioSpan;
};

export type StudioImage = {
    pageNumber: number;
    confidence?: Confidence;
    span: StudioSpan;
};

export type StudioLine = {
    content: string;
    polygon: Polygon;
    spans: StudioSpan[];
};

export type StudioBlock = {
    kind: string;
    BoundingRegion: BoundingRegion[];
    content: string;
    spans: StudioSpan[];
    confidence: Confidence;
};

export type StudioParagraph = {
    spans: StudioSpan[];
    boundingRegions?: BoundingRegion[];
    content: string;
    role?: string;
};

export type StudioSelectionMark = {
    state: SelectionMarkState;
    polygon: Polygon;
    confidence?: Confidence;
    span?: StudioSpan;
};

export type StudioDocumentTableCell = {
    kind?: DocumentTableCellKind;
    rowIndex: number;
    columnIndex: number;
    rowSpan?: number;
    columnSpan?: number;
    content: any;
    boundingRegions: BoundingRegion[];
    spans?: StudioSpan[];
};

type StudioKeyValuePairElement = {
    content: string;
    boundingRegions: BoundingRegion[];
    spans: StudioSpan[];
};

export type StudioFieldValue = {
    type: FieldValueType;
    valueString?: string;
    valueDate?: string;
    valueTime?: string;
    valuePhoneNumber?: string;
    valueNumber?: number;
    valueInteger?: number;
    valueSelectionMark?: SelectionMarkState;
    valueSignature?: DocumentSignatureType;
    valueCountryRegion?: string;
    valueCurrency?: DocumentCurrencyType;
    valueArray?: StudioFieldValue[];
    valueObject?: { [name: string]: StudioFieldValue };
    valueAddress?: { [name: string]: string };
    content?: string;
    boundingRegions?: BoundingRegion[];
    confidence?: Confidence;
    spans?: StudioSpan[];
};

export type StudioSpan = {
    offset: number;
    length: number;
};

export type Polygon = [number, number, number, number, number, number, number, number];

type DocumentCurrencyType = {
    currencySymbol?: string;
    currency?: string;
    amount?: number;
};

export type SelectionMarkState = "selected" | "unselected";

type Confidence = number;

type DocumentTableCellKind = "content" | "rowHeader" | "columnHeader" | "stub" | "description";

type FieldValueType =
    | "string"
    | "date"
    | "time"
    | "phoneNumber"
    | "number"
    | "integer"
    | "selectionMark"
    | "countryRegion"
    | "currency"
    | "signature"
    | "array"
    | "object"
    | "address";

type DocumentSignatureType = "signed" | "unsigned";

export type SpanPositions = {
    offset: number;
    length: number;
    iconName: string;
    pageNumber?: number;
    tableIndex?: number;
};

export type AnalyzeContentParseResult = {
    documentName: string;
    content: string;
    displayedContent: string;
    positions: Array<{
        column: number;
        lineNumber: number;
        iconName: string;
        shouldDisplayAsIcon: boolean;
        pageNumber?: number;
        tableIndex?: number;
    }>;
};

export type ParsedContentTextBlock = {
    kind?: string | undefined;
    role?: string | undefined;
    content: string;
    boundingRegions?: BoundingRegion[];
};

export type ParsedContentPagedText = {
    [pageNumber: string]: ParsedContentTextBlock[];
};

export type ParsedContentPagedTables = {
    [pageNumber: string]: StudioDocumentTable[];
};

export type ParsedContentPagedSelectionMarks = {
    [pageNumber: string]: StudioSelectionMark[];
};

export type StudioGeneralKeyValue = StudioDocumentKeyValuePair & {
    fieldColor?: string;
};

export type StudioGeneralEntityValue = {
    category: string;
    fieldColor: string;
    valueArray: StudioDocumentEntity[];
};
