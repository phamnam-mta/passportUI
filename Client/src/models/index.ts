export type SplitPaneSizes = {
    analyzeSplitPaneSize: number[];
    labelSplitPaneSize: number[];
    labelTableSplitPaneSize: number[];
};
