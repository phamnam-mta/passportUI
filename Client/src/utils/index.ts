export const delay = (ms: number) => {
    return new Promise<void>((resolve) => {
        setTimeout(() => {
            resolve();
        }, ms);
    });
};

export const getPixelWidthFromPercent = (widthInPercentage: number): number => {
    const totalWindowWidth = document.documentElement.clientWidth;
    const paneWidthInPixel = (totalWindowWidth / 100) * widthInPercentage;
    return paneWidthInPixel;
};

export const loadCanvasToBlob = (canvas: HTMLCanvasElement): Promise<Blob> => {
    return new Promise((resolve, reject) => {
        canvas.toBlob((blob) => {
            if (blob) {
                resolve(blob);
            } else {
                reject("Failed to create blob for canvas.");
            }
        });
    });
};

export const loadUrlToArrayBuffer = async (url: string): Promise<ArrayBuffer> => {
    return await fetch(url).then((r) => r.arrayBuffer());
};

export const debounce = (func, timeout = 300) => {
    return (...args) => {
        return setTimeout(() => {
            func.apply(this, args);
        }, timeout);
    };
};
