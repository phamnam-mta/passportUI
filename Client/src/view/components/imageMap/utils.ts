import { Feature } from "ol";
import { Extent } from "ol/extent";
import Polygon from "ol/geom/Polygon";
import ImageLayer from "ol/layer/Image";
import Layer from "ol/layer/Layer";
import VectorLayer from "ol/layer/Vector";
import Projection from "ol/proj/Projection";
import ImageSource from "ol/source/ImageStatic";
import VectorSource from "ol/source/Vector";
import { LAYER_NAME } from "./constants";
import { IRegion } from "./contracts";

export type Dimension = {
    width: number;
    height: number;
};

export const degreeToRadians = (degree: number) => {
    // convert degree to radians
    return (degree * Math.PI * 2) / 360;
};

export const makeLayerFilter = (layerName) => {
    return {
        layerFilter: (layer: Layer) => layer.get(LAYER_NAME) === layerName,
    };
};

export const makeVectorLayer = (layerName, options = {}): Layer => {
    return new VectorLayer({
        [LAYER_NAME]: layerName,
        source: new VectorSource(),
        ...options,
    } as any);
};

export const makeImageLayer = (layerName, uri, projection, imageExtent, options = {}): Layer => {
    return new ImageLayer({
        [LAYER_NAME]: layerName,
        source: makeImageSource(uri, projection, imageExtent),
        ...options,
    } as any);
};

export const makeImageSource = (url: string, projection: Projection, imageExtent: Extent) => {
    return new ImageSource({ url, projection, imageExtent });
};

export const convertRegionToFeature = (region: IRegion, imageExtent: Extent, isOcrProposal: boolean = false) => {
    const coordinates: Array<[number, number]> = [];
    const boundingBox = region.id.split(",").map(parseFloat);
    const imageWidth = imageExtent[2] - imageExtent[0];
    const imageHeight = imageExtent[3] - imageExtent[1];
    for (let i = 0; i < boundingBox.length; i += 2) {
        coordinates.push([Math.round(boundingBox[i] * imageWidth), Math.round((1 - boundingBox[i + 1]) * imageHeight)]);
    }
    const feature = new Feature({
        geometry: new Polygon([coordinates]),
    });
    feature.setProperties({
        id: region.id,
        text: region.value,
        highlighted: false,
        isOcrProposal,
    });
    feature.setId(region.id);

    return feature;
};

export const createRegionIdFromPolygon = (polygon: number[], page: number): string => {
    return polygon.join(",") + ":" + page;
};

export const convertToImageMapCoordinates = (polygon: number[], canvasSize: Dimension, documentSize: Dimension) => {
    const { width: canvasWidth, height: canvasHeight } = canvasSize;
    const { width: ocrWidth, height: ocrHeight } = documentSize;

    const coordinates: number[][] = [];
    for (let i = 0; i < polygon.length; i += 2) {
        coordinates.push([(polygon[i] / ocrWidth) * canvasWidth, (1 - polygon[i + 1] / ocrHeight) * canvasHeight]);
    }
    return coordinates;
};

export const getBoundingBoxFromFeatureId = (id) => id.split(",").map(parseFloat);
