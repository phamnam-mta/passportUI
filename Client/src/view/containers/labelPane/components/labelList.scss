@import "~@fluentui/react/dist/sass/References";

.label-list {
    overflow: auto;
    height: 100%;
    .item-separator {
        width: 100%;
        border: 0;
        border-bottom: 1px solid $ms-color-gray30;
        margin: 0;
    }
}
