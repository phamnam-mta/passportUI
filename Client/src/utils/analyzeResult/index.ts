import { BoundingRegion, Polygon, StudioFieldValue, StudioSpan } from "models/analyzeResult";

export const getPagePolygons = (boundingRegions: BoundingRegion[], page: number): Polygon[] => {
    return boundingRegions
        .filter((boundingRegion) => boundingRegion.pageNumber === page)
        .map((boundingRegion) => boundingRegion.polygon);
};

export const isAnyBoundingRegionInPage = (boundingRegions: BoundingRegion[], page: number): boolean => {
    return boundingRegions.some((boundingRegion) => boundingRegion.pageNumber === page);
};

export const getFieldTypeValue = (field: StudioFieldValue) => {
    const { type } = field;
    const typeString = type.charAt(0).toUpperCase() + type.slice(1);

    // TODO: Update when going public.
    if (type === "currency" && field.valueCurrency && typeof field.valueCurrency !== "string") {
        const { currencySymbol, currency, amount } = field.valueCurrency;
        if (currencySymbol && amount) {
            return `${currencySymbol}${amount}`;
        } else if (currency && amount) {
            return `${currency}${amount}`;
        } else if (amount) {
            return `${amount}`;
        } else {
            return undefined;
        }
    } else if (type === "address" && field.valueAddress) {
        return field.content;
    }
    return field[`value${typeString}`];
};

export interface SpanPosition extends StudioSpan {
    iconName: string;
    pageNumber?: number;
    tableIndex?: number;
    shouldDisplayAsIcon: boolean;
}
