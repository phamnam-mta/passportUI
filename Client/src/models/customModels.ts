import { FeatureCategory } from "view/components/imageMap/contracts";
import { Polygon } from "./analyzeResult";

export type CustomModel = {
    modelId: string;
    description: string;
    createdDateTime: string;
    apiVersion?: string;
    tags?: {
        modelType?: string;
        [key: string]: string | undefined;
    };
};

export type PrimitiveField = {
    fieldKey: string;
    fieldType: FieldType;
    fieldFormat: FieldFormat;
};

export type ObjectField = {
    fieldKey: string;
    fieldType: FieldType;
    fields: Field[];
    fieldFormat: FieldFormat;
    visualizationHint?: VisualizationHint;
};

export type ArrayField = {
    fieldKey: string;
    fieldType: FieldType;
    itemType: string;
    visualizationHint?: VisualizationHint;
};

export type Field = PrimitiveField | ObjectField | ArrayField;

export type FieldsWithOrder = Field & { order: number };

export enum FieldType {
    String = "string",
    Number = "number",
    Date = "date",
    Time = "time",
    Integer = "integer",
    SelectionMark = "selectionMark",
    Signature = "signature",
    Array = "array",
    Object = "object",
}

export enum FieldFormat {
    NotSpecified = "not-specified",
    Currency = "currency",
    Decimal = "decimal",
    DecimalCommaSeparated = "decimal-comma-separated",
    NoWhiteSpaces = "no-white-spaces",
    Alphanumeric = "alphanumeric",
    DMY = "dmy",
    MDY = "mdy",
    YMD = "ymd",
}

export enum VisualizationHint {
    Horizontal = "horizontal",
    Vertical = "vertical",
}

export type Labels = {
    [documentName: string]: Label[];
};

export type Label = {
    label: string;
    value: LabelValue[];
    labelType?: LabelType;
};

export type LabelValue = {
    boundingBoxes: Polygon[];
    page: number;
    text: string;
};

export type LabelValueCandidate = {
    boundingBoxes: Polygon[];
    page: number;
    text: string;
    category: FeatureCategory;
    alreadyAssignedLabelName?: string;
};

export enum LabelType {
    Words = "words",
    Region = "region",
}

export type Definitions = {
    [objectName: string]: ObjectField;
};

export enum TableType {
    dynamic = "dynamic",
    fixed = "fixed",
}

export enum HeaderType {
    row = "row",
    column = "column",
}
