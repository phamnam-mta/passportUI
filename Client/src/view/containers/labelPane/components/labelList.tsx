import React from "react";
import { Field, FieldType, Label } from "models/customModels";
import LabelItem from "./labelItem";
import { getColorByFieldKey, getFieldColor, getFieldKeyFromLabel, getUnusedFieldColor } from "utils/customModel";
import { connect, ConnectedProps } from "react-redux";
import { ApplicationState } from "store";
import { setHoveredLabelName } from "store/canvas/canvas";
import {
    assignLabel,
    deleteField,
    deleteLabelByField,
    renameField,
    setColorForFields,
    setHideInlineLabelMenu,
    switchSubType,
    updateFieldsOrder,
} from "store/customModel/customModel";
import {
    DragDropContext,
    Droppable,
    Draggable,
    DroppableProvided,
    DraggableProvided,
    DropResult,
} from "react-beautiful-dnd";
import update from "immutability-helper";

import "./labelList.scss";

interface ILabelListProps {
    onTablePaneOpen: (field: Field) => void;
}

export class LabelList extends React.PureComponent<ILabelListProps & ConnectedProps<typeof connector>> {
    public componentDidUpdate(prevProps) {
        const { fields, colorForFields, setColorForFields } = this.props;

        if (fields !== prevProps.fields) {
            if (fields.length > prevProps.fields.length) {
                const addedFields = fields.filter((field) => !prevProps.fields.includes(field));
                const addedColorMap = addedFields.map((field) => ({
                    [field.fieldKey]:
                        prevProps.fields.length === 0
                            ? getFieldColor(fields, field.fieldKey)
                            : getUnusedFieldColor(colorForFields),
                }));

                setColorForFields([...colorForFields, ...addedColorMap]);
            }
            if (fields.length < prevProps.fields.length) {
                const removedFields = prevProps.fields.filter((field) => !fields.includes(field));
                const removedKeys = removedFields.map((field) => field.fieldKey);

                setColorForFields(colorForFields.filter((color) => Object.keys(color)[0] !== removedKeys[0]));
            }
        }
    }

    private getDocumentLabels = (): Label[] => {
        const { currentDocument, labels } = this.props;
        if (currentDocument) {
            return labels[currentDocument.name] || [];
        } else {
            return [];
        }
    };

    private handleDeleteLabel = async (fieldKey: string) => {
        const { deleteLabelByField } = this.props;
        await deleteLabelByField(fieldKey);
    };

    private handleSwitchSubType = (fieldKey: string, fieldType: FieldType) => {
        const { switchSubType } = this.props;
        switchSubType({ fieldKey, fieldType });
    };

    private handleDeleteField = async (deletingField: string) => {
        const { deleteField } = this.props;
        await deleteField(deletingField);
    };

    private handleRenameField = async (oldFieldName: string, newFieldName: string) => {
        const { renameField } = this.props;
        await renameField({ fieldKey: oldFieldName, newName: newFieldName });
    };

    private onGetCreateFieldErrorMessage = (value: string) => {
        const { fields } = this.props;
        const isDuplicate = fields.some((field) => field.fieldKey === value);

        if (isDuplicate) {
            return "The field already exists.";
        } else {
            return undefined;
        }
    };

    private handleAssignLabel = (labelName: string) => {
        const { assignLabel } = this.props;
        assignLabel(labelName);
    };

    private handleTablePaneOpen = (field: Field) => {
        const { setHideInlineLabelMenu, onTablePaneOpen } = this.props;
        setHideInlineLabelMenu(true);
        onTablePaneOpen(field);
    };

    private handleItemMouseEnter = (labelName: string) => {
        const { hoveredLabelName, setHoveredLabelName } = this.props;
        if (hoveredLabelName !== labelName) {
            setHoveredLabelName(labelName);
        }
    };

    private handleItemMouseLeave = () => {
        this.props.setHoveredLabelName("");
    };

    private handleDragEnd = (result: DropResult) => {
        if (!result.destination) {
            return;
        }

        const destinationIndex = result.destination.index;
        const sourceIndex = result.source.index;
        const { updateFieldsOrder, fields } = this.props;

        if (destinationIndex === sourceIndex) {
            return;
        }

        updateFieldsOrder(
            update(fields, {
                $splice: [
                    [sourceIndex, 1],
                    [destinationIndex, 0, fields[sourceIndex]],
                ],
            })
        );
    };

    public render() {
        const { fields, colorForFields } = this.props;
        return (
            <div className="label-list">
                <DragDropContext onDragEnd={this.handleDragEnd}>
                    <Droppable droppableId="droppable">
                        {(droppableProvided: DroppableProvided) => (
                            <div ref={droppableProvided.innerRef} {...droppableProvided.droppableProps}>
                                {fields.map((field, index) => {
                                    // Table will not use label but will still pass this in for checking label existence.
                                    const { fieldKey } = field;
                                    const label = this.getDocumentLabels().find(
                                        (label) => getFieldKeyFromLabel(label) === fieldKey
                                    );
                                    return (
                                        <div key={fieldKey}>
                                            <Draggable key={fieldKey} draggableId={fieldKey} index={index}>
                                                {(draggableProvided: DraggableProvided) => (
                                                    <div
                                                        ref={draggableProvided.innerRef}
                                                        {...draggableProvided.draggableProps}
                                                    >
                                                        {index !== 0 && <hr className="item-separator" />}
                                                        <LabelItem
                                                            field={field}
                                                            label={label}
                                                            color={getColorByFieldKey(colorForFields, fieldKey)}
                                                            onDeleteLabel={this.handleDeleteLabel}
                                                            onSwitchSubType={this.handleSwitchSubType}
                                                            onDeleteField={this.handleDeleteField}
                                                            onRenameField={this.handleRenameField}
                                                            onGetRenameErrorMessage={this.onGetCreateFieldErrorMessage}
                                                            onClickTableField={this.handleTablePaneOpen}
                                                            onClickField={this.handleAssignLabel}
                                                            onItemMouseEnter={this.handleItemMouseEnter}
                                                            onItemMouseLeave={this.handleItemMouseLeave}
                                                            dragHandleProps={draggableProvided.dragHandleProps}
                                                        />
                                                        {draggableProvided.placeholder}
                                                    </div>
                                                )}
                                            </Draggable>
                                        </div>
                                    );
                                })}
                                {droppableProvided.placeholder}
                            </div>
                        )}
                    </Droppable>
                </DragDropContext>
            </div>
        );
    }
}

const mapState = (state: ApplicationState) => ({
    currentDocument: state.documents.currentDocument,
    fields: state.customModel.fields,
    colorForFields: state.customModel.colorForFields,
    labels: state.customModel.labels,
    hoveredLabelName: state.canvas.hoveredLabelName,
});
const mapDispatch = {
    switchSubType,
    deleteField,
    renameField,
    deleteLabelByField,
    assignLabel,
    setHoveredLabelName,
    setHideInlineLabelMenu,
    updateFieldsOrder,
    setColorForFields,
};

const connector = connect(mapState, mapDispatch);

export default connector(LabelList);
