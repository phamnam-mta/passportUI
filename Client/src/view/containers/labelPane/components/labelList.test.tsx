import * as React from "react";
import { LabelList } from "./labelList";
import { FieldType } from "models/customModels";
import { mockDocument, mockDocumentLabels, mockFields, mockDynamicTableField, mockColorForFields } from "utils/test";
import { DropResult } from "react-beautiful-dnd";
import update from "immutability-helper";
import { shallow } from "enzyme";

describe("<LabelList />", () => {
    let baseProps;
    let droppableProvided;
    let draggableProvided;
    let getLabelItem;

    const mockFieldKey = "mock key";

    beforeEach(() => {
        baseProps = {
            onTablePaneOpen: jest.fn(),
            currentDocument: mockDocument,
            fields: mockFields,
            labels: {
                [mockDocument.name]: mockDocumentLabels,
            },
            hoveredLabelName: "",
            colorForFields: mockColorForFields,
            switchSubType: jest.fn(),
            deleteField: jest.fn(),
            renameField: jest.fn(),
            deleteLabelByField: jest.fn(),
            assignLabel: jest.fn(),
            setHideInlineLabelMenu: jest.fn(),
            setHoveredLabelName: jest.fn(),
            updateFieldsOrder: jest.fn(),
            setColorForFields: jest.fn(),
        };
        droppableProvided = { innerRef: "", droppableProps: [] };
        draggableProvided = { innerRef: "", draggableProps: [], dragHandleProps: [] };

        getLabelItem = (index = 0) => {
            const wrapper = shallow(<LabelList {...baseProps} />);
            const droppable = wrapper.find("Connect(Droppable)");
            const droppableInner = shallow((droppable.prop("children") as any)(droppableProvided, null));
            const draggable = droppableInner.find("PublicDraggable").at(index);
            const draggableInner = shallow((draggable.prop("children") as any)(draggableProvided, null));
            return draggableInner.find("Connect(LabelItem)");
        };
    });

    describe("Rendering", () => {
        it("should render correctly", () => {
            const wrapper = shallow(<LabelList {...baseProps} />);
            const droppable = wrapper.find("Connect(Droppable)");
            const droppableInner = shallow((droppable.prop("children") as any)(droppableProvided, null));
            const draggable = droppableInner.find("PublicDraggable").first();
            const draggableInner = shallow((draggable.prop("children") as any)(draggableProvided, null));

            expect(wrapper).toMatchSnapshot();
            expect(droppableInner).toMatchSnapshot();
            expect(draggableInner).toMatchSnapshot();
        });

        it("should match snapshot, when there's no fields", () => {
            const props = { ...baseProps, fields: [], labels: {} };
            const wrapper = shallow(<LabelList {...props} />);

            const droppable = wrapper.find("Connect(Droppable)");
            const droppableInner = shallow((droppable.prop("children") as any)(droppableProvided, null));

            expect(wrapper).toMatchSnapshot();
            expect(droppableInner).toMatchSnapshot();
        });
    });

    describe("Event Handling", () => {
        it("should handle open table pane, when table field is clicked", () => {
            const labelItem = getLabelItem();
            labelItem.prop("onClickTableField")(mockDynamicTableField);

            expect(baseProps.setHideInlineLabelMenu).toBeCalledTimes(1);
            expect(baseProps.setHideInlineLabelMenu).toBeCalledWith(true);
            expect(baseProps.onTablePaneOpen).toBeCalledTimes(1);
            expect(baseProps.onTablePaneOpen).toBeCalledWith(mockDynamicTableField);
        });

        it("should handle delete label, when LabelItem onDeleteLabel is triggered", () => {
            const labelItem = getLabelItem();
            labelItem.prop("onDeleteLabel")(mockFieldKey);

            expect(baseProps.deleteLabelByField).toBeCalledTimes(1);
            expect(baseProps.deleteLabelByField).toBeCalledWith(mockFieldKey);
        });

        it("should handle switch sub type, when LabelItem onSwitchSubType is triggered", () => {
            const mockFieldType = FieldType.Number;

            const labelItem = getLabelItem();
            labelItem.prop("onSwitchSubType")(mockFieldKey, mockFieldType);

            expect(baseProps.switchSubType).toBeCalledTimes(1);
            expect(baseProps.switchSubType).toBeCalledWith({
                fieldKey: mockFieldKey,
                fieldType: mockFieldType,
            });
        });

        it("should handle delete field, when LabelItem onDeleteField is triggered", () => {
            const labelItem = getLabelItem();
            labelItem.prop("onDeleteField")(mockFieldKey);

            expect(baseProps.deleteField).toBeCalledTimes(1);
            expect(baseProps.deleteField).toBeCalledWith(mockFieldKey);
        });

        it("should handle rename field, when LabelItem onRenameField is triggered", () => {
            const newName = "newFieldName";

            const labelItem = getLabelItem();
            labelItem.prop("onRenameField")(mockFieldKey, newName);

            expect(baseProps.renameField).toBeCalledTimes(1);
            expect(baseProps.renameField).toBeCalledWith({ fieldKey: mockFieldKey, newName });
        });

        it("should return exist error, when LabelItem onGetRenameErrorMessage is triggered with existing field", () => {
            const labelItem = getLabelItem();
            const error = labelItem.prop("onGetRenameErrorMessage")(mockFields[0].fieldKey);

            expect(error).toBe("The field already exists.");
        });

        it("should not return error, when LabelItem onGetRenameErrorMessage is triggered with new field", () => {
            const labelItem = getLabelItem();
            const error = labelItem.prop("onGetRenameErrorMessage")("Non-existing-field");

            expect(error).toBeUndefined();
        });

        it("should handle assign label, when LabelItem onClickField is triggered", () => {
            const labelItem = getLabelItem();
            const labelName = "label~1name";
            labelItem.prop("onClickField")(labelName);

            expect(baseProps.assignLabel).toBeCalledTimes(1);
            expect(baseProps.assignLabel).toBeCalledWith(labelName);
        });

        it("should drag and drop handles properly", () => {
            const wrapper = shallow(<LabelList {...baseProps} />) as any;

            const getWrapperItem = (index) => {
                const droppable = wrapper.find("Connect(Droppable)");
                const droppableInner = shallow((droppable.prop("children") as any)(droppableProvided, null));
                const draggable = droppableInner.find("PublicDraggable").at(index);
                const draggableInner = shallow((draggable.prop("children") as any)(draggableProvided, null));
                return draggableInner.find("Connect(LabelItem)");
            };

            expect(getWrapperItem(0).prop("field")).toBe(mockFields[0]);
            expect(getWrapperItem(1).prop("field")).toBe(mockFields[1]);

            const dragAndDropContext = wrapper.find("DragDropContext") as any;
            const sourceIndex = 0;
            const destinationIndex = 1;
            const dragEndResult: DropResult = {
                destination: {
                    index: destinationIndex,
                },
                source: {
                    index: sourceIndex,
                },
            };

            const switchedFields = update(mockFields, {
                $splice: [
                    [sourceIndex, 1],
                    [destinationIndex, 0, mockFields[sourceIndex]],
                ],
            });
            dragAndDropContext.prop("onDragEnd")(dragEndResult);

            expect(baseProps.updateFieldsOrder).toBeCalledTimes(1);
            expect(baseProps.updateFieldsOrder).toBeCalledWith(switchedFields);

            wrapper.setProps({
                fields: switchedFields,
            });

            expect(getWrapperItem(0).prop("field")).toBe(mockFields[1]);
            expect(getWrapperItem(1).prop("field")).toBe(mockFields[0]);
        });
    });
});
