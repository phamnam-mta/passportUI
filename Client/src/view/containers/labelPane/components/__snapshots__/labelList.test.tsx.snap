// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LabelList /> Rendering should match snapshot, when there's no fields 1`] = `
<div
  className="label-list"
>
  <DragDropContext
    onDragEnd={[Function]}
  >
    <Connect(Droppable)
      direction="vertical"
      droppableId="droppable"
      getContainerForClone={[Function]}
      ignoreContainerClipping={false}
      isCombineEnabled={false}
      isDropDisabled={false}
      mode="standard"
      renderClone={null}
      type="DEFAULT"
    >
      <Component />
    </Connect(Droppable)>
  </DragDropContext>
</div>
`;

exports[`<LabelList /> Rendering should match snapshot, when there's no fields 2`] = `<div />`;

exports[`<LabelList /> Rendering should render correctly 1`] = `
<div
  className="label-list"
>
  <DragDropContext
    onDragEnd={[Function]}
  >
    <Connect(Droppable)
      direction="vertical"
      droppableId="droppable"
      getContainerForClone={[Function]}
      ignoreContainerClipping={false}
      isCombineEnabled={false}
      isDropDisabled={false}
      mode="standard"
      renderClone={null}
      type="DEFAULT"
    >
      <Component />
    </Connect(Droppable)>
  </DragDropContext>
</div>
`;

exports[`<LabelList /> Rendering should render correctly 2`] = `
<div>
  <div
    key="Address"
  >
    <PublicDraggable
      draggableId="Address"
      index={0}
      key="Address"
    >
      <Component />
    </PublicDraggable>
  </div>
  <div
    key="SelectionMark"
  >
    <PublicDraggable
      draggableId="SelectionMark"
      index={1}
      key="SelectionMark"
    >
      <Component />
    </PublicDraggable>
  </div>
  <div
    key="Signature"
  >
    <PublicDraggable
      draggableId="Signature"
      index={2}
      key="Signature"
    >
      <Component />
    </PublicDraggable>
  </div>
  <div
    key="DynamicTable/abc"
  >
    <PublicDraggable
      draggableId="DynamicTable/abc"
      index={3}
      key="DynamicTable/abc"
    >
      <Component />
    </PublicDraggable>
  </div>
  <div
    key="FixedRowTable"
  >
    <PublicDraggable
      draggableId="FixedRowTable"
      index={4}
      key="FixedRowTable"
    >
      <Component />
    </PublicDraggable>
  </div>
  <div
    key="FixedColumnTable"
  >
    <PublicDraggable
      draggableId="FixedColumnTable"
      index={5}
      key="FixedColumnTable"
    >
      <Component />
    </PublicDraggable>
  </div>
</div>
`;

exports[`<LabelList /> Rendering should render correctly 3`] = `
<div>
  <Connect(LabelItem)
    color="#a0aeb2"
    dragHandleProps={Array []}
    field={
      Object {
        "fieldFormat": "not-specified",
        "fieldKey": "Address",
        "fieldType": "string",
      }
    }
    label={
      Object {
        "label": "Address",
        "value": Array [
          Object {
            "boundingBoxes": Array [
              Array [
                0.8573176470588235,
                0.1557818181818182,
                0.8605529411764706,
                0.1557818181818182,
                0.8605529411764706,
                0.1626909090909091,
                0.8573176470588235,
                0.1626909090909091,
              ],
            ],
            "page": 1,
            "text": "NEW",
          },
          Object {
            "boundingBoxes": Array [
              Array [
                0.8705411764705883,
                0.1557818181818182,
                0.8902823529411765,
                0.1557818181818182,
                0.8902823529411765,
                0.1628,
                0.8705411764705883,
                0.1628,
              ],
            ],
            "page": 1,
            "text": "BELGIUM",
          },
          Object {
            "boundingBoxes": Array [
              Array [
                0.8986117647058824,
                0.1557818181818182,
                0.918364705882353,
                0.1557818181818182,
                0.918364705882353,
                0.1628,
                0.8986117647058824,
                0.1628,
              ],
            ],
            "page": 1,
            "text": "BREWERY",
          },
        ],
      }
    }
    onClickField={[Function]}
    onClickTableField={[Function]}
    onDeleteField={[Function]}
    onDeleteLabel={[Function]}
    onGetRenameErrorMessage={[Function]}
    onItemMouseEnter={[Function]}
    onItemMouseLeave={[Function]}
    onRenameField={[Function]}
    onSwitchSubType={[Function]}
  />
</div>
`;
